File: androidApp/src/main/java/com/brettm/holiday/android/UIState.kt
package com.brettm.holiday.android

sealed class UIState<out Content : Any> {
    data class Loading<Content : Any>(val content: Content?) : UIState<Content>()
    data class Loaded<Content : Any>(val content: Content) : UIState<Content>()
    data class Error(val error: String) : UIState<Nothing>()

    fun content(): Content? {
        return when (this) {
            is Loading -> content
            is Loaded -> content
            else -> null
        }
    }
}

File: iosApp/iosApp/UIState.swift
import Foundation

enum UIState<Content> {
    case loading(Content?)
    case loaded(Content)
    case error(String)

    func content() -> Content? {
        switch self {
        case .loading(let content): return content
        case .loaded(let content): return content
        default: return nil
        }
    }
}
